### PR TITLE
Randomly play opening moves based on priors

### LINKED ISSUE #342: Separate exploration from training feedback (alternate method of lowering T)
We have two competing goals: 
A) Get an accurate value of positions.
B) Explore different positions so we learn about a variety of different positions.
For B we have T=1 for the entire game. But this reduces the accuracy of A.

Proposed method:
1) Pick a random number of plies N
2) For N plies, do one playout. Pick the move to play proportional to the Policy output.
  2a) For these moves do not record anything in the training data file
3) After N plies, do our normal self play algorithm with T=0. Note that Dirichlet noise will still introduce some amount of a randomness.

For picking N we want a good number of games that start near the opening so we can learn values of opening moves. But we also want to get into unusual middle/endgame positions. 

Using this method, the time spent in step 1 is wasted in that we generate no data to feedback to our NN. But this step is 800 times faster because we only do one playout for each move in that step.

### PR DESCRIPTION
**Pull request by Mardak**
r?@Tilps or @mooskagh Fixes #342 by adding selfplay `--random-opening-max-plies` that results in a random number of plies doing a single visit to randomly pick a move proportional to the (noised) priors while not saving training data for these early moves. Can also change `--random-opening-temperature` from 1.0 to larger/flatten or smaller/sharper priors.

Not explicitly mentioned in the issue but implemented here is the skipping of adjudication during this random opening to hopefully allow more opening diversity of positions that otherwise would have been overlooked if assuming opponent will play best move -> resign.

This doesn't change the existing default behavior of 0 prior-based random moves, but it allows the server to change it and even in addition to regular temperature settings.

If the game ends while playing randomly, it's treated like an abort.

This includes/supersedes #773 as I needed to move the `MakeMove` earlier anyway (but still after getting training data).

### CODE REVIEW

**Comment by Mardak** [comment]
As a sanity check letting the random openings play the random number of plies up to 30 then immediately adjudicating:
```
./lc0 selfplay -w 53200 --visits=1 --games=10000 --resign-percentage=100 --max-random-opening-plies=30

# 326 games that had 0-ply openings (followed by highest prior move to end the game)
 326 e2e4

# 300 games that had 1-ply openings (followed by highest prior move to end the game)
   1 a2a4 g8f6
   1 e2e3 d7d5
   2 b2b3 e7e5
   2 b2b4 e7e5
   3 a2a4 e7e5
   4 b1c3 d7d5
   4 h2h3 e7e5
   5 f2f4 g8f6
   5 g2g4 d7d5
   5 h2h4 e7e5
   6 b1a3 e7e5
   7 e2e3 e7e6
   7 f2f3 e7e5
   7 g1h3 d7d5
   7 g2g3 e7e5
   8 a2a3 e7e5
   8 c2c3 g8f6
   8 c2c4 e7e5
   9 d2d3 d7d5
  10 d2d4 g8f6
  12 g1f3 g8f6
 179 e2e4 e7e6

# 318 games that had 2-ply openings (followed by highest prior move to end the game)
…
   1 h2h4 g8h6 e2e4
   2 a2a4 d7d5 g1f3
   2 b1a3 e7e5 a3c4
   2 b2b3 e7e6 c1b2
   2 b2b4 e7e5 c1b2
   2 c2c4 c7c5 g1f3
   2 d2d3 d7d5 g1f3
   2 e2e3 c7c5 d2d4
   2 e2e3 e7e6 g1f3
   2 e2e4 b7b5 f1b5
   2 e2e4 e7e6 d2d4
   2 e2e4 g8f6 e4e5
   3 a2a3 e7e5 c2c4
   3 c2c4 e7e5 g2g3
   3 d2d4 d7d5 c2c4
   3 e2e4 g8h6 d2d4
   3 e2e4 h7h6 d2d4
   3 f2f4 g8f6 e2e3
   3 g1f3 d7d5 d2d4
   3 g1f3 g8f6 d2d4
   4 d2d4 g8f6 g1f3
   4 e2e4 b8a6 g1f3
   4 e2e4 f7f5 e4f5
   4 e2e4 f7f6 d2d4
   4 e2e4 h7h5 d2d4
   4 g1h3 d7d5 g2g3
   6 b1c3 d7d5 d2d4
   6 e2e4 d7d6 d2d4
   6 h2h3 e7e5 c2c4
   7 e2e4 b7b6 d2d4
   7 e2e4 b8c6 d2d4
   8 e2e4 a7a6 d2d4
   8 e2e4 d7d5 e4d5
   9 e2e4 c7c6 d2d4
  11 e2e4 g7g6 d2d4
  16 e2e4 e7e5 g1f3
  25 e2e4 c7c5 g1f3
  81 e2e4 e7e6 b1c3
```

Those total per-ply counts do match up pretty closely to 1 / (30 + 1) * 10000 ~= 323

**Comment by Tilps** [comment]
I think we'll want a temperature parameter for this pre-training period. And maybe some other parameters. But I am favorable to the concept.

**Comment by Mardak** [comment]
I can understand the desire to have additional parameters like temperature, but do we have a concrete example of what behavior we want more than controlling the max? And even then, should that be done in a followup?

And to be clear, those existing move temperature options can be used after the random opening, e.g., if random opening picked 0-ply, regular visit-ful search can still use the 6 existing options: Temperature, TempDecayMoves, TempCutoffMove, TempEndgame, TempValueCutoff and TempVisitOffset.

**Comment by aartdappel** [comment]
adding a temp to this seems key, at least T40 has very extreme policies in main lines where this wouldn't help exploration at all

**Comment by Mardak** [comment]
The priors used for the random opening are after noise has been applied to a newly extended root node. So with something like #267, there would be a way to control the priors a bit instead of the fixed 25% epsilon -- although that would affect noise everywhere and not just the opening…

**Comment by aartdappel** [comment]
It does seem easier to me to just increase the temp value than to do a temp over the noise. That way you don't have to tune noise for this specifically. Having two sources of randomness interact with each other so directly doesn't seem beneficial

**Comment by Tilps** [comment]
Noise will certainly help here - maybe that is good enough - but I expect some additional flattening would be quite likely to be useful.  We already use temp > 1 for our first moves - while they have visits and this does not, it does seem intuitive that the temperature used before the first moves of the game should be even higher.

**Comment by Mardak** [comment]
Updated PR with two options: `--random-opening-max-plies` and `--random-opening-temperature` (the latter can be set per player1. vs player2.)

**Comment by Mardak** [comment]
Another sanity check testing temperature for first ply using 53200 with 10k games:

```
--random-opening-temperature=100
 460 h2h4
 461 a2a3
 463 b1c3
 464 c2c4
 468 b2b4
 486 g2g4
 491 f2f3
 492 b1a3
 496 g1f3
 496 g1h3
 498 f2f4
 501 c2c3
 503 d2d4
 504 e2e3
 532 a2a4
 532 e2e4
 536 g2g3
 536 h2h3
 538 b2b3
 543 d2d3

--random-opening-temperature=10
 450 g1h3
 455 f2f3
 459 f2f4
 460 g2g3
 470 b2b3
 473 d2d3
 475 h2h4
 476 g2g4
 479 b1c3
 482 a2a4
 484 b2b4
 488 c2c3
 490 a2a3
 494 b1a3
 507 h2h3
 515 d2d4
 531 c2c4
 536 e2e3
 544 g1f3
 732 e2e4

--random-opening-temperature=1
 134 g2g4
 148 g1h3
 149 a2a3
 149 b1a3
 153 h2h3
 154 h2h4
 170 a2a4
 173 f2f3
 179 b1c3
 181 f2f4
 183 b2b4
 191 b2b3
 196 d2d3
 210 c2c3
 220 e2e3
 239 c2c4
 257 g2g3
 328 g1f3
 329 d2d4
6257 e2e4

--random-opening-temperature=0.5
  10 g2g4
  11 f2f3
  12 a2a3
  13 h2h3
  14 d2d3
  16 b2b4
  17 c2c3
  17 f2f4
  19 a2a4
  19 g1h3
  21 b1a3
  22 h2h4
  23 b2b3
  23 e2e3
  28 c2c4
  29 g2g3
  37 b1c3
  39 d2d4
  44 g1f3
9586 e2e4

--random-opening-temperature=0.1
10000 e2e4
```

And similarly except forcing e2e4 as the opening move:
```
--random-opening-temperature=100
 453 e2e4 b8a6
 460 e2e4 d7d5
 466 e2e4 h7h6
 472 e2e4 f7f6
 475 e2e4 a7a6
 493 e2e4 g8f6
 495 e2e4 g8h6
 496 e2e4 c7c6
 500 e2e4 h7h5
 502 e2e4 b7b6
 507 e2e4 b8c6
 507 e2e4 g7g6
 511 e2e4 f7f5
 512 e2e4 g7g5
 518 e2e4 a7a5
 518 e2e4 c7c5
 523 e2e4 b7b5
 524 e2e4 d7d6
 525 e2e4 e7e6
 543 e2e4 e7e5

--random-opening-temperature=10
 412 e2e4 h7h5
 421 e2e4 f7f5
 439 e2e4 g8h6
 447 e2e4 f7f6
 454 e2e4 g8f6
 455 e2e4 b7b6
 465 e2e4 b7b5
 466 e2e4 b8a6
 470 e2e4 g7g5
 478 e2e4 d7d5
 488 e2e4 d7d6
 491 e2e4 b8c6
 493 e2e4 h7h6
 496 e2e4 a7a5
 517 e2e4 a7a6
 539 e2e4 g7g6
 559 e2e4 c7c6
 596 e2e4 e7e5
 620 e2e4 c7c5
 694 e2e4 e7e6

--random-opening-temperature=1
 136 e2e4 g7g5
 138 e2e4 g8h6
 146 e2e4 b7b5
 153 e2e4 f7f5
 155 e2e4 a7a5
 166 e2e4 h7h6
 171 e2e4 b8a6
 174 e2e4 h7h5
 179 e2e4 f7f6
 196 e2e4 b7b6
 202 e2e4 d7d5
 231 e2e4 g8f6
 283 e2e4 g7g6
 284 e2e4 d7d6
 301 e2e4 b8c6
 332 e2e4 a7a6
 501 e2e4 c7c6
 774 e2e4 e7e5
1561 e2e4 c7c5
3917 e2e4 e7e6

--random-opening-temperature=0.5
  29 e2e4 a7a5
  31 e2e4 b7b5
  31 e2e4 h7h5
  34 e2e4 b8a6
  34 e2e4 d7d5
  36 e2e4 g7g5
  37 e2e4 f7f6
  37 e2e4 g8h6
  43 e2e4 f7f5
  43 e2e4 g8f6
  45 e2e4 h7h6
  46 e2e4 b7b6
  58 e2e4 g7g6
  66 e2e4 d7d6
  83 e2e4 b8c6
  84 e2e4 a7a6
 143 e2e4 c7c6
 332 e2e4 e7e5
1042 e2e4 c7c5
7746 e2e4 e7e6

--random-opening-temperature=0.1
10000 e2e4 e7e6
```

**Comment by Mardak** [comment]
@Tilps I rebased on latest master resolving adjacent-line conflicts with #821. Is the current set of options sufficient for the server to control the behavior at least for a first attempt:
- random-opening-temperature
- random-opening-max-plies

**Comment by Naphthalin** [comment]
I just wanted to point out that the matter discussed in this PR is within the expected behavior of using #918 in training with ``temp=1.0`` where the % of each line is given by the equlibrium policies.

**Comment by Mardak** [comment]
#342 was fixed with #964 so closing this.
